src/main/java/bio/terra/workspace/db/DataReferenceDao.java

```diff
@@ -0,0 +1,120 @@
+    this.objectMapper = objectMapper;
+  }
+
+  public String createDataReference(
+      UUID referenceId,
+      UUID workspaceId,
+      String name,
+      JsonNullable<UUID> resourceId,
+      JsonNullable<String> credentialId,
+      String cloningInstructions,
+      JsonNullable<DataReference.ReferenceTypeEnum> referenceType,
+      JsonNullable<DataRepoSnapshot> reference) {
+    String sql =
+        "INSERT INTO workspace_data_reference (workspace_id, reference_id, name, resource_id, credential_id, cloning_instructions, reference_type, reference) VALUES "
+            + "(:workspace_id, :reference_id, :name, :resource_id, :credential_id, :cloning_instructions, :reference_type, cast(:reference AS json))";
+
+    Map<String, Object> paramMap = new HashMap<>();
+    paramMap.put("workspace_id", workspaceId.toString());
+    paramMap.put("reference_id", referenceId.toString());
+    paramMap.put("name", name);
+    paramMap.put("cloning_instructions", cloningInstructions);
+
+    if (credentialId.isPresent()) {
+      paramMap.put("credential_id", credentialId.get());
+    } else paramMap.put("credential_id", null);
+
+    if (resourceId.isPresent()) {
+      paramMap.put("resource_id", resourceId.get().toString());
+      paramMap.put("reference_type", null);
+      paramMap.put("reference", null);
+    } else {
+      if (referenceType.isPresent() && reference.isPresent()) {
+        paramMap.put("resource_id", null);
+        paramMap.put("reference_type", referenceType.get().getValue());
+        try {
+          paramMap.put("reference", objectMapper.writeValueAsString(reference));
+        } catch (JsonProcessingException e) {
+          throw new InvalidDataReferenceException("Could not write data reference to database");
+        }
+      } else
+        throw new InvalidDataReferenceException(
+            "DataReference must contain either a resource ID or a reference type and a reference");
+    }
+
+    jdbcTemplate.update(sql, paramMap);
+    return referenceId.toString();
+  }
+
+  public DataReference getDataReference(UUID referenceId) {
+    String sql =
+        "SELECT workspace_id, reference_id, name, resource_id, credential_id, cloning_instructions, reference_type, reference from workspace_data_reference where reference_id = :id";
+
+    Map<String, Object> paramMap = new HashMap<>();
+    paramMap.put("id", referenceId.toString());
+
+    Map<String, Object> queryOutput = jdbcTemplate.queryForMap(sql, paramMap);
+
+    DataReference reference = new DataReference();
+
+    reference.setWorkspaceId(UUID.fromString(queryOutput.get("workspace_id").toString()));
+    reference.setReferenceId(UUID.fromString(queryOutput.get("reference_id").toString()));
+    reference.setName(queryOutput.get("name").toString());
+    reference.setReferenceType(
+        JsonNullable.of(
+            DataReference.ReferenceTypeEnum.fromValue(
+                queryOutput.get("reference_type").toString())));
```

Comment:
style nit: I just recently found out OpenApi generated objects have builder-style methods, so you can do something like
```
DataReference reference = new DataReference()
.workspaceId(...)
.referenceId(...)
...;
```
Mostly a style thing, so your call on doing that or not.